extensions/answer_summarizers/models.py

```diff
@@ -227,36 +225,45 @@ def calculate_from_state_answers_dict(self, state_answers_dict):
 
         This method is run from within the context of a MapReduce job.
         """
-        top_answer_counts_as_list_of_pairs = _count_answers(
-            state_answers_dict['submitted_answer_list'])
+        classify_categories = [
+            exp_domain.EXPLICIT_CLASSIFICATION,
+            exp_domain.TRAINING_DATA_CLASSIFICATION,
+            exp_domain.STATISTICAL_CLASSIFICATION,
+            exp_domain.DEFAULT_OUTCOME_CLASSIFICATION
+        ]
+
+        submitted_answer_list = state_answers_dict['submitted_answer_list']
+        submitted_answers_by_categorization = {
+            classify_category: [
+                submitted_answer_dict
+                for submitted_answer_dict in submitted_answer_list
+                if submitted_answer_dict['classification_categorization'] == (
+                    classify_category)]
+            for classify_category in classify_categories
+        }
+        top_answer_count_pairs_by_category = {
+            classify_category: _count_answers(answers)
+            for classify_category, answers
+            in submitted_answers_by_categorization.iteritems()
+        }
 
         calculation_output = {
-            exp_domain.EXPLICIT_CLASSIFICATION: [],
-            exp_domain.TRAINING_DATA_CLASSIFICATION: [],
-            exp_domain.STATISTICAL_CLASSIFICATION: [],
-            exp_domain.DEFAULT_OUTCOME_CLASSIFICATION: [],
+            classify_category: []
+            for classify_category in classify_categories
         }
-        for item in top_answer_counts_as_list_of_pairs:
-            answer_dict = item[0]
-            classify_category = answer_dict['classification_categorization']
-            if classify_category not in calculation_output:
-                raise Exception(
-                    'Cannot aggregate answer with unknown rule classification '
-                    'category: %s' % classify_category)
-            calculation_output[classify_category].append({
-                'answer': answer_dict['answer'],
-                'frequency': item[1]
-            })
+        for classify_category, top_answer_counts_as_list_of_pairs in (
```

Comment:
Might be worth changing the data structure in order to avoid the awkward 'list of pairs' formulation...